### ISSUE TITLE
Use of EscapeUriString in URL formation leads to docs that can be created but not deleted

### ISSUE THREAD

**Issue by abhijitpai** [Azure/azure-cosmos-dotnet-v3#2970]
**Describe the bug**
https://docs.microsoft.com/en-us/dotnet/api/system.uri.escapeuristring?view=net-6.0
EscapeUriString (obsolete in recent .net versions) is used in https://github.com/Azure/azure-cosmos-dotnet-v3/blob/d331683a3937cc435002f291cd1998f8c5b93c46/Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.Items.cs#L1207 and similar to form URLs that doesn't encode parameters correctly. These should instead use EscapeDataString.

**To Reproduce**
Direct mode:
Item item1 = new Item() { Id = "abcd/", Partitionkey = "pk1" };
await container.CreateItemAsync(item1, new PartitionKey(item1.Partitionkey));
await container.DeleteItemAsync<Item>("abcd/", new PartitionKey(item1.Partitionkey));

**Expected behavior**
Both operations succeed.

**Actual behavior**
Create succeeds but delete throws 404 because the service is sent the document name "abcd" without any trace of "/".

**Environment summary**
SDK Version: 3.23.0
OS Version (e.g. Windows, Linux, MacOSX): Windows.

**Additional context**
Add any other context about the problem here (for example, complete stack traces or logs).

**Comment by ealsur** [comment]
@abhijitpai We used to have `/` documented as an unsupported character for ids? How does the REST API handle reads for items with `/` on the Uri when using named routes?

**Comment by abhijitpai** [comment]
It is likely that our REST API implementation does not handle it correctly. But by sticking to the obsolete method, we are having users need to find other much costlier workarounds after they create such documents, so I don't see why this should not be fixed.

**Comment by ealsur** [comment]
Agree that if the method is obsolete, it should be addressed. The only thing is we don't target NET 6, but NET Standard 2.0, and it's not obsolete there:

![image](https://user-images.githubusercontent.com/1633401/154723327-614c99ff-3641-42ba-a670-d35a0fdc28fa.png)

**Comment by abhijitpai** [comment]
We are talking about reducing customer pain here.

**Comment by abhijitpai** [comment]
It is relevant due to the fact it is made obsolete because it doesn't encode properly, but anyway we can let that be. REST API is not in my control, and neither is it in yours - but the SDK is. I am not sure why it matters that we need to fix the REST API to resolve the customer problem when we can make this fix in the SDK.

**Comment by ealsur** [comment]
@abhijitpai I changed the method in the code and did a simple test:

```
CosmosClient cosmosClient = new CosmosClient("<connectionstring>", new CosmosClientOptions() { ConnectionMode = ConnectionMode.Gateway });

            var container = cosmosClient.GetContainer("test", "items");
            var item1 = new { Id = "abcd/", Partitionkey = "pk1" };
            await container.CreateItemAsync(item1, new PartitionKey(item1.Partitionkey));
            await container.DeleteItemAsync<dynamic>("abcd/", new PartitionKey(item1.Partitionkey));
```

Gateway returns a 401:

![image](https://user-images.githubusercontent.com/1633401/154726312-c783b2ff-cdd9-4c9e-aecb-ae9f703faa93.png)

![image](https://user-images.githubusercontent.com/1633401/154726463-6ff32c78-1c56-451c-9e2c-cb37a2d9fd2b.png)

![image](https://user-images.githubusercontent.com/1633401/154726562-40b146c1-d794-4cbb-afeb-ab8cf1f30b2d.png)

![image](https://user-images.githubusercontent.com/1633401/154726638-35d15477-1531-4150-a5eb-94100da84114.png)


![image](https://user-images.githubusercontent.com/1633401/154727640-6a037222-e904-4f99-9303-de52f73f0f87.png)

![image](https://user-images.githubusercontent.com/1633401/154727883-563ce132-8fa5-4ea7-9a93-9ffd79834b15.png)

The problem is way deeper than just the `EscapeDataString`, it is deep in the PathsHelper, the code does this:

![image](https://user-images.githubusercontent.com/1633401/154729171-f009ae88-4729-4133-9be8-8ace274d5ce7.png)

It calls `Uri.UnescapeDataString` which basically removes the escaping:

![image](https://user-images.githubusercontent.com/1633401/154729247-0ed50a6b-59d8-4d22-acfa-656303a12209.png)

Looks like in order to support this scenario, it will also require changes in the PathHelper and Resource parsing.

**Comment by abhijitpai** [comment]
Please try it in direct mode.

**Comment by ealsur** [comment]
Direct mode yields a 404 even after hacking the code to have the EscapeDataString and avoid the unescaping on the PathHelper.

![image](https://user-images.githubusercontent.com/1633401/154732589-91e90d46-e367-43f1-b1c0-1784c11b31a8.png)

![image](https://user-images.githubusercontent.com/1633401/154732630-4def61b2-3de4-4dad-8ed0-ff1433e4ff37.png)

Could it be because really the document's `id` is not `abcd%2F` but `abcd/`?

In Direct mode, if I don't do any changes, the document is deleted correctly. So the scenario currently works in Direct mode.

**Comment by ealsur** [comment]
The escaping would only be needed in Gateway mode, in which case, the Gateway itself is not supporting it.

**Comment by abhijitpai** [comment]
Yes, the DocumentName you have above is incorrect - it should match the user specified document name which is "abcd/". And that also explains how we address both the documents in your last image - for the second document, the DocumentName that needs to be passed is "abcd%2F".
